### PR TITLE
Expand detail around the integer values of enum types

### PR DESCRIPTION
**Pull request by ckreibich**
As per today's [discussion on Slack](https://zeekorg.slack.com/archives/CTGDG83B8/p1649694398602929).

@bbannier, @ynadji, please feel free to take a look, thanks.

### CODE REVIEW

File: script-reference/types.rst [hunk 1]
```diff
@@ -788,11 +788,29 @@ have no further structure.  An example declaration:
...
+    type color: enum { Red = 10, White = 20, Blue = 30 };
+
+Without explicit assignment, Zeek numbers enumerations sequentially starting
+from 0. You may not mix explicit and implicit assignment.
+
+The only operations allowed on enumerations are equality comparisons (``==``,
+``!=``) and assignment (``=``). Enumerations do not automatically yield their
+values or provide ordering (neither ``Red == 10`` nor ``Red < White`` works),
+but the :zeek:see:`enum_to_int` BIF lets you retrieve an enumeration's numeric
+values if you require such logic.
```
Comment (bbannier) [inline script-reference/types.rst:807]:
```suggestion
value if you require such logic.
```

File: script-reference/types.rst [hunk 2]
```diff
@@ -788,11 +788,29 @@ have no further structure.  An example declaration:
...
+
+.. note::
+
+   We recommend using explicit value assignment when relying on numeric values,
+   since it avoids surprises when enumerations are :zeek:keyword:`redef`'d in
+   multiple scripts.
```
Comment (bbannier) [inline script-reference/types.rst:813]: _Surprises_ hints that there is potential for sinister stuff to happen, maybe instead spell out the benefits, e.g., something like

```suggestion
   We recommend using explicit value assignment when relying on numeric values
   to provide defined semantics for conversions to integers with :zeek:keyword:`enum_to_int`.
```

File: script-reference/types.rst [hunk 3]
```diff
@@ -788,11 +788,29 @@ have no further structure.  An example declaration:
...
 
 The last comma after ``Blue`` is optional.  Both the type name ``color``
-and the individual values (``Red``, etc.) have global scope.
+and the individual values (``Red``, etc. -- not ``color::Red``) have
+global scope.
 
-Enumerations do not have associated values or ordering.
-The only operations allowed on enumerations are equality comparisons
-(``==``, ``!=``) and assignment (``=``).
+Enumerations may assign :zeek:type:`count` values explicitly:
```
Comment (bbannier) [inline script-reference/types.rst:794]: Note: there's not test for this currently.

**Review by bbannier — APPROVED** [review]

File: script-reference/types.rst (hunk 2 above)
Comment (ckreibich) [reply script-reference/types.rst]: Thanks! I've reworded in the subsequent force-push to spell out the problem.

File: script-reference/types.rst (hunk 3 above)
Comment (ynadji) [reply script-reference/types.rst:794]: This test has them explicitly assigned: https://github.com/zeek/zeek/blob/master/testing/btest/bifs/enum_to_int.zeek

**Review by ynadji — APPROVED** [review]
LGTM!

File: script-reference/types.rst (hunk 3 above)
Comment (bbannier) [reply script-reference/types.rst:794]: Nothing in the BTest baselines currently has the error string if one uses e.g., a negative integer (`enumerator is not a count constant`). This is hard to test for in general though 🤷 .

Comment (ynadji) [reply script-reference/types.rst:794]: Ahh I see the distinction now, thanks

**Pull request merged**